### ISSUE TITLE
Add the option to invert the raster color ramp values in the Layer Styling panel

### ISSUE THREAD

**Issue by stuskier** [qgis/QGIS#41341]
**Feature description.**

At 3.16.3, raster color ramps are only displayed with the *lowest* raster value at the top of the display list.  However, there are many times when the *highest* value is required at the top of the list.  That is, the values are displayed from high to low, top to bottom.  

It would be very useful if a "switch" was available in the Layer Styling panel that allowed the user to invert the display of the value order.

A full description, with screenshots, can be seen at:

https://gis.stackexchange.com/questions/385525/qgis-how-to-invert-the-values-in-a-layout-legend

**Additional context**

**Comment by rduivenvoorde** [comment]
It's not only raster, the same for vector styles if I am correct.

**Comment by stuskier** [comment]
Hello rduivenvoorde, you are correct - it is for both vector and raster.  I have edited the original request.  Thanks!

**Comment by benwirf** [comment]
I have been playing around with this in my own build and have added a tool button to the layout legend widget to invert a layer's legend nodes. Although this feature request mentions a control via the layer styling, the linked gis.stackexchange question is about inverting the values in a _**layout legend**_ which, fwiw, I think is the better approach. This would be fairly tidy, contained addition to the codebase, allowing users to simply uncheck 'Auto update' and quickly invert color ramp nodes in the layout legend as a complement to the 'Move up' and 'Move down' tool buttons. I think this would cover the majority of use cases, but I'm very happy to hear comments or thoughts on the pros/cons or shortcomings of this approach.
Please a demonstration gif below. The tool button is only enabled upon selection of a legend layer which has child nodes and works with both raster and vector layers.

![Invert_legend_nodes](https://user-images.githubusercontent.com/47767794/108188733-a63e5e80-715b-11eb-957c-fdb88a6ca649.gif)

**Comment by gioman** [comment]
> but I'm very happy to hear comments or thoughts on the pros/cons or shortcomings of this approach.

@benwirf you may want to create a pull request in order to get the maximum attention and feedback possible from the developers.

**Comment by benwirf** [comment]
Thanks @gioman. I will tidy up the code a bit & make a PR.

**Comment by stuskier** [comment]
@benwirf Thanks for taking on this topic. Your gif was informative, which I note also sorts alphabetically in addition to numerically.  I was originally thinking only numeric sorting, but you went beyond that.  

I have a comment on your proposal:  I think that the sorting should be enabled both at the QGIS main map view (via the Layer Styling panel) as well as the layout view.  My original SE post title used the phrase "layout legend", but, in retrospect, I should not have limited it to just the layout legend.  There are plenty of times I make a map without going to the layout view, and for those cases, being able to re-sort within the Layer Styling panel would be very important.  Sorry for the confusion that I may have caused; I can edit the OP if you prefer.

**Comment by benwirf** [comment]
@stuskier, no problem. I am just starting to get into C++ QGIS development so taking on small features like this is good practice for me.
Just to clarify, there is no actual sorting going on. The logic behind this is simply reversing the current legend node order rather than making comparisons of string or numeric values, so it works the same way regardless of the legend value datatype.
As far as extending this to the main QGIS map view, My personal view is that implementing this via the layer properties symbology dialog or styling dock would require more time and effort and would also have to touch a lot more of the code base due to the multiple renderer types which result in child legend nodes for both vector and raster layers.
I propose another option to just modify the legend in the layers panel either via adding a tool button to the layers panel dock widget or adding an action to the context menu when right-clicking a legend item which has child nodes.

<img width="363" alt="layer_panel_tool_button" src="https://user-images.githubusercontent.com/47767794/108579018-cf840800-7370-11eb-8624-1cf2fdaecd94.png">

<img width="393" alt="context_menu" src="https://user-images.githubusercontent.com/47767794/108579028-d90d7000-7370-11eb-8856-7ccca64bdb8a.png">

Would this approach satisfy your use case? I am happy to work on implementing this as well as in my spare time but I would probably split it into two separate pull request and just reference this feature request in both.

**Comment by rduivenvoorde** [comment]
@benwirf mmm, I thought too look into where to put some button/menu in the layer styling dialog and panel...
Then found out... that you can actually just sort the color ramp....
Well, at least for categorized and graduated... For rule-based it does NOT seem to work...

![Screenshot-20210220163446-300x448](https://user-images.githubusercontent.com/731673/108600887-da9f5e00-7399-11eb-9f62-50cbd59d02ac.png)

![Screenshot-20210220163434-348x465](https://user-images.githubusercontent.com/731673/108600886-d70bd700-7399-11eb-9e4e-192347e7f790.png)

Did you know? (I did not...)If you can get your way somewhere in an intuitive place in the Layer Styling panels, that would be ok too?

**Comment by stuskier** [comment]
@rduivenvoorde that is a great find!  There it was, right in front of me all the time.  

I did a little research, and this is what I found:

### Vector data:
Within the **Layer Styling** panel, sorting can be done on _Categorized, Graduated_, and _Inverted Polygon_ styles.  Additionally, this sorting automatically transfers to a Layout legend in real time; thus, the Layout is always synchronized to the main map.

Sorting is not available with _Rule-based_ styling.

### Raster data:
Sorting is not available at all.

So, @benwirf perhaps you could use your coding skills to add sorting to Rule-based vector data, and/or to add sorting to raster data?

**Comment by benwirf** [comment]
@rduivenvoorde, I did find that option to sort the table columns in the categorized renderer however I was initially sorting the 'Legend' column and I found that it sorts like this:

![Screenshot from 2021-02-21 09-38-28](https://user-images.githubusercontent.com/47767794/108621605-52938400-747f-11eb-9733-5085b6ad007c.png)

E.g. **1-19 then 2** etc. which I saw as a limitation, but after further checking, I found that, as you showed in your screenshots, if you sort the 'Value' column it works perfectly!
Additionally, there is also an option for rasters- if you use a singleband-pseudocolor renderer with a linear interpolation, you can click the 'Legend Settings' button and select either 'Maximum on top' or Minimum on the top'.

![Screenshot from 2021-02-21 09-41-37](https://user-images.githubusercontent.com/47767794/108621892-3b559600-7481-11eb-8e59-589b246b39be.png)

Which results in a legend item where the color ramp is a single node:

![Screenshot from 2021-02-21 09-41-57](https://user-images.githubusercontent.com/47767794/108621941-84a5e580-7481-11eb-9190-4fb9e6780e9f.png)

However this doesn't seem to be available with e.g. paletted renderer or discrete interpolation.
So, in summary there are already ways to sort _some_ color ramps in the various symbology dialogs, however they are not very consistent and do not work in every situation. 
I'm thinking aloud (and I could be wrong) here but the major difference I see is that, in the symbology dialogs, the ordering is done in a Table View/Model which I'm pretty sure will use a true sorting algorithm. This could make it difficult to properly sort columns with values containing characters such as '<=' or '>'. Whereas, in the legend itself which a Tree View/ QgsLayerTreeModel which is based on QAbstractItemModel, there is a very handy way to set the legend layer node order using QgsMapLayerLegendUtils.
@stuskier, I will have a look into sorting the columns in the other raster renderer widgets, but tbh I'm not hugely motivated to work on these at the moment. I still kind of think that a single tool button or context menu action in the TOC panel would be a nice clean solution! :-) For now I will keep an open mind and keep exploring both possibilities.

**Comment by stuskier** [comment]
@benwirf I agree with everything you said.  Thanks, also, for you honesty about your motivation level, and good luck!  In closing, what QGIS version are you using?  I'm at 3.16.3, and I don't see the raster _Layer Properties > Legend _Settings__ button that you show.  Here's what my screen looks like:

![Capture](https://user-images.githubusercontent.com/12805900/108652640-7eac0680-7479-11eb-94dc-0468e3aaf6a5.PNG)

**Comment by benwirf** [comment]
@stuskier, No problem- cheers. Legend settings button was in master version, so it should be available now in 3.18!
